README.md
# fileStreamer

[![GitHub release](https://img.shields.io/github/v/release/DigiLive/fileStreamer?include_prereleases)](https://github.com/DigiLive/gitChangelog/releases)
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

This library serves a file according to the headers which are sent with a http
request. It supports resumable downloads or streaming the content of a file to a
client.

If you have any questions, comments or ideas concerning this library, please
consult the code documentation at first.
Create a new [issue](https://github.com/DigiLive/fileStreamer/issues/new) if
your concerns remain unanswered.

## Features

* Inline disposition.
* Attachment disposition.
* Serve a complete file.
* Serve a single byte range of a file.
* Serve multiple byte ranges of a file.
* Set a custom mimetype.

## Requirements

* PHP ^7.4
* ext-fileinfo *

## Installation

The preferred method is to install the library
with [Composer](http://getcomposer.org).

```sh
> composer require digilive/file-streamer:^1
```

Set the version constraint to a value which suits you best.  
Alternatively you can download the latest release
from [GitHub](https://github.com/DigiLive/fileStreamer/releases).

## Example use

```php
<?php

use DigiLive\FileStreamer\FileStreamer;

// Use composer's auto loader.
$requiredFile = 'Path/To/vendor/autoload.php';

// Or include the library manually.
// $requiredFile = 'Path/To/FileStreamer.php';

require_once $requiredFile;

// Instantiate the library.
$fileStreamer = new FileStreamer('Path/To/File/To/Serve.ext');
// Set inline disposition if wished.
$fileStreamer->setInline();
$fileStreamer->start();
// Execution of PHP will terminate when FileStreamer::start() is finished.
```


CHANGELOG.md
# Changelog

## v1.0.0 (2022-10-02)

* Add setting a custom mimetype ([79e3079][0])

[0]:https://github.com/DigiLive/fileStreamer/commit/79e3079


tests/FileStreamerTest.php
<?php /** @noinspection PhpUndefinedFieldInspection */

/** @noinspection PhpUndefinedMethodInspection */

declare(strict_types=1);

namespace DigiLive\FileStreamer\Tests;

use DigiLive\FileStreamer\FileStreamer;
use DigiLive\Output\Output;
use DigiLive\Reflector\Reflector;
use Exception;
use org\bovigo\vfs\vfsStream;
use org\bovigo\vfs\vfsStreamException;
use org\bovigo\vfs\vfsStreamWrapper;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;

/**
 * PHPUnit test for class DigiLive\FileStreamer.
 *
 * PHP version 7.4 or greater
 *
 * @package         DigiLive\FileStreamer\Tests
 * @author          Ferry Cools <info@DigiLive.nl>
 * @copyright   (c) 2022 Ferry Cools
 * @license         New BSD License http://www.opensource.org/licenses/bsd-license.php
 * @version         1.0.0
 * @link            https://github.com/DigiLive/fileStreamer
 */
class FileStreamerTest extends TestCase
{
    /**
     * Defines the headers which are always sent by the FileStreamer class.
     * Some could be overridden at a real download, but the Output class will append instead of override.
     */
    private const defaultHeaders = [
        'Pragma: public',
        'Expires: -1',
        'Cache-Control: public, must-revalidate, post-check=0, pre-check=0',
        'Accept-Ranges: bytes',
        'Content-Type: text/plain',                                  // Could be overridden at real download.
        'Content-Transfer-Encoding: binary',
        'Content-Disposition: attachment; filename="dummyFile.txt"', // Could be overridden at real download.
        'Content-Length: 10',                                        // Could be overridden at real download.
    ];

    /**
     * Instantiate mocks required for this test case.
     *
     * Mocks include a filesystem and an output buffer.
     *
     * @return void
     */
    public static function setUpBeforeClass(): void
    {
        vfsStream::setup('source', null, ['dummyFile.txt' => '0123456789']);
        Output::getOverrides('./supportFiles/OutputFunctions.php');
        Output::reset();
    }

    /**
     * Clear mocks after the last test of this test class is run.
     *
     * @throws vfsStreamException When unregistering the mocked filesystem fails.
     */
    public static function tearDownAfterClass(): void
    {
        vfsStreamWrapper::unregister();
        Output::reset();
    }

    /**
     * Reset the collected output and headers after each test.
     *
     * @return void
     */
    public function tearDown(): void
    {
        Output::reset();
    }

    /**
     * Test if range values get extracted from the ranges as defined in the http request header.
     *
     * @dataProvider provideValidRanges
     *
     * @param   string  $RequestedRange  A valid range.
     * @param   array   $expectedRange   Expected range extracted from the valid range.
     *
     * @return void
     */
    public function testGetRequestedRanges(string $RequestedRange, array $expectedRange)
    {
        $_SERVER['HTTP_RANGE'] = $RequestedRange;
        $fileStreamer          = new FileStreamer(vfsStream::url('source/dummyFile.txt'));
        $reflector             = Reflector::create($fileStreamer);

        $reflector->getRequestedRanges();

        $this->assertEquals($expectedRange, $reflector->fileRanges);
        $reflector->unset();
    }

    /**
     * Test if the correct headers are set when requesting an invalid range.
     *
     * @dataProvider provideInvalidRanges
     *
     * @param   string  $RequestedRange  An invalid range.
     *
     * @return void
     */
    public function testRequestedRangesInvalid(string $RequestedRange)
    {
        $_SERVER['HTTP_RANGE'] = $RequestedRange;

        // Mock the class to override the terminate method.
        /** @var FileStreamer $mock */
        $mock      = $this->mockFileStreamer();
        $reflector = Reflector::create($mock);

        try {
            $reflector->getRequestedRanges();
            $this->fail();
        } catch (Exception $e) {
            $this->assertCount(1, Output::$headers);
            $this->assertContains('HTTP/1.1 416 Requested Range Not Satisfiable', Output::$headers);
        }

        $reflector->unset();
    }

    /**
     * Mock the FileStreamer class to override the terminate method.
     *
     * The terminate method contains an exit statement which will abort the execution of php.
     * To prevent this, this method is overridden and will only raise an exception instead.
     *
     * @return MockObject The mocked FileDownload class.
     */
    private function mockFileStreamer(): MockObject
    {
        $mock = $this->getMockBuilder(FileStreamer::class)
                     ->onlyMethods(['terminate'])
                     ->setConstructorArgs([vfsStream::url('source/dummyFile.txt')])
                     ->getMock();
        $mock->method('terminate')->willThrowException(new Exception('Mocked'));

        return $mock;
    }

    /**
     * Test if the correct data is being served according to the requested range.
     *
     * @dataProvider provideValidRanges
     *
     * @param   string  $RequestedRange   A valid range.
     * @param   array   $expectedRange    Unused.
     * @param   string  $expectedOutput   Expected content which is sent to the outputbuffer.
     * @param   array   $expectedHeaders  Expected headers which are sent to the outputbuffer.
     *
     * @return void
     */
    public function testStartSingleRange(
        string $RequestedRange,
        array $expectedRange,
        string $expectedOutput,
        array $expectedHeaders
    ): void {
        // Skip multi-range dataset.
        if ($this->dataName() == 'multiRange') {
            $this->markTestSkipped('Test skipped.');
        }

        $_SERVER['HTTP_RANGE'] = $RequestedRange;
        $expectedHeaders       = array_merge(self::defaultHeaders, $expectedHeaders);

        // Get a mocked instance of the class with overridden terminate method.
        /** @var FileStreamer $mock */
        $mock = $this->mockFileStreamer();

        try {
            // Suppress warning or method will be aborted.
            @$mock->start();
            $this->fail();
        } catch (Exception $e) {
            $this->assertSame($expectedOutput, Output::getBody());
            $this->assertSame($expectedHeaders, Output::$headers);
        }
    }

    /**
     * Test if the correct data is being served according to the requested ranges.
     *
     * @return void
     */
    public function testStartMultiRange(): void
    {
        $_SERVER['HTTP_RANGE'] = 'bytes=2-3,5-6,-1';
        $expectedHeaders       = [
            'HTTP/1.1 206 Partial Content',
            'Content-Length: 330',
            'Content-Type: multipart/byteranges; boundary=cafd507eba83d389029d38c0cbe92dc5',
        ];
        $expectedHeaders       = array_merge(self::defaultHeaders, $expectedHeaders);
        $expectedOutput        = <<<'TXT'

--cafd507eba83d389029d38c0cbe92dc5
Content-Type: text/plain
Content-range: bytes 2-3/10

23
--cafd507eba83d389029d38c0cbe92dc5
Content-Type: text/plain
Content-range: bytes 5-6/10

56
--cafd507eba83d389029d38c0cbe92dc5
Content-Type: text/plain
Content-range: bytes 9-9/10

9
--cafd507eba83d389029d38c0cbe92dc5--

TXT;
        // Ensure the line separators match \r\n.
        $expectedOutput = preg_replace('~\R~u', "\r\n", $expectedOutput);

        // Get a mocked instance of the class with overridden terminate method.
        /** @var FileStreamer $mock */
        $mock = $this->mockFileStreamer();

        try {
            // Suppress warning or method will be aborted.
            @$mock->start();
            $this->fail();
        } catch (Exception $e) {
            $this->assertSame($expectedOutput, Output::getBody());
            $this->assertSame($expectedHeaders, Output::$headers);
        }
    }

    /**
     * Test if the correct data is being served when no range is requested.
     * Also, it tests setting a custom Content-Type.
     *
     * @return void
     */
    public function testStartNoRange(): void
    {
        unset($_SERVER['HTTP_RANGE']);
        // Get a mocked instance of the class with overridden terminate method.
        /** @var FileStreamer $mock */
        $mock = $this->mockFileStreamer();

        // Attachment disposition Test.
        try {
            // Suppress warning or method will be aborted.
            @$mock->start();
        } catch (Exception $e) {
            // Test headers.
            $this->assertSame(self::defaultHeaders, Output::$headers);
        }

        // Inline disposition Test.
        Output::reset();
        $mock->setInline(true);
        $mock->setMimeType('digilive/test');

        $expectedHeaders = array_merge(self::defaultHeaders, ['Content-Disposition: inline']);
        $expectedHeaders[4] = 'Content-Type: digilive/test';

        try {
            // Suppress warning or method will be aborted.
            @$mock->start();
        } catch (Exception $e) {
            // Test headers.
            $this->assertSame($expectedHeaders, Output::$headers);
            $this->assertSame('0123456789', Output::getBody());
        }
    }

    /**
     * Valid ranges are:
     * bytes=0-500                  // The first 500 bytes.
     * bytes=-500                   // The last 500 bytes, not 0-500!
     * bytes=500-                   // From byte 500 tot the end.
     * bytes=0-500,1000-1499,-200   // The first 500 bytes, From byte 1000 to 1499 and the last 200 bytes.
     *
     * @return array[] Valid ranges and the expected start/end bytes extracted from these ranges.
     */
    public function provideValidRanges(): array
    {
        return [
            'singleRange' => [
                'bytes=3-7',
                [['start' => 3, 'end' => 7]],
                '34567',
                [
                    'HTTP/1.1 206 Partial Content',
                    'Content-Length: 5',
                    'Content-Range: bytes 3-7/10',
                ],
            ],
            'endRange'    => [
                'bytes=-3',
                [['start' => 7, 'end' => 9]],
                '789',
                [
                    'HTTP/1.1 206 Partial Content',
                    'Content-Length: 3',
                    'Content-Range: bytes 7-9/10',
                ],
            ],
            'startRange'  => [
                'bytes=3-',
                [['start' => 3, 'end' => 9]],
                '3456789',
                [
                    'HTTP/1.1 206 Partial Content',
                    'Content-Length: 7',
                    'Content-Range: bytes 3-9/10',
                ],
            ],
            'multiRange'  => [
                'bytes=2-4,5-7,-2',
                [
                    ['start' => 2, 'end' => 4],
                    ['start' => 5, 'end' => 7],
                    ['start' => 8, 'end' => 9],
                ],
                '23456789',
                [],
            ],
        ];
    }

    /**
     * Invalid ranges are:
     * - Unit isn't bytes.
     * - Start is greater than end.
     *
     * @return string[][] Invalid ranges.
     */
    public function provideInvalidRanges(): array
    {
        return [
            'invalidUnit'        => [
                'invalid=3-7',
            ],
            'invalidStartSingle' => [
                'bytes=9-7',
            ],
            'invalidStartMulti'  => [
                'bytes=2-4,7-5,-2',
            ],
        ];
    }
}


tests/supportFiles/OutputFunctions.php
<?php
/**
 * Functions overrides for the DigiLive\Output package.
 *
 * Headers and output which would normally be sent to php's output buffer, are caught in this class.
 *
 * PHP version 7.4 or greater
 *
 * @package         DigiLive\Output
 * @author          Ferry Cools <info@DigiLive.nl>
 * @copyright   (c) 2022 Ferry Cools
 * @version         1.0.0
 * @license         New BSD License http://www.opensource.org/licenses/bsd-license.php
 * @todo            Get this package from composer.
 */
declare(strict_types=1);

namespace DigiLive\FileStreamer;

// Above namespace must be set the same as the namespace which calls the defined functions below.

use DigiLive\Output\Output;

/**
 * headers_sent will return false if no HTTP headers have already been sent or true otherwise.
 *
 * @return false
 * @see \headers_sent()
 */
function headers_sent(): bool
{
    return false;
}

/**
 * Capture a raw HTTP header.
 *
 * Note:
 * Unlike the default header function, a captured similar header will not be replaced.
 * A new header is captured instead.
 *
 * @param   string  $value  The header string.
 *
 * @return void
 * @see \header()
 */
function header(string $value)
{
    Output::$headers[] = $value;
}

/**
 * Capture a formatted string.
 *
 * @param   string  $format
 * @param   mixed   ...$values
 *
 * @return void
 * @see \printf()
 */
function printf(string $format, ...$values)
{
    Output::$body .= sprintf($format, ...$values);
}

/**
 * Capture the body of the output buffer.
 *
 * @return void
 * @see \ob_get_contents();
 */
function ob_flush(): void
{
    Output::$body .= ob_get_contents();
    ob_clean();
}

/**
 * Flush system output buffer.
 *
 * Since the output buffer is captured, try to flush this buffer should do noting.
 *
 * @return void
 * @see \flush()
 */
function flush(): void
{
}


tests/supportFiles/DigiLive/Output/Output.php
<?php

declare(strict_types=1);

namespace DigiLive\Output;

/**
 * Mimic the output buffer.
 *
 * Headers and output which would normally be sent to php's output buffer, are caught in this class.
 *
 * PHP version 7.4 or greater
 *
 * @package         DigiLive\Output
 * @author          Ferry Cools <info@DigiLive.nl>
 * @copyright   (c) 2022 Ferry Cools
 * @version         1.0.0
 * @license         New BSD License http://www.opensource.org/licenses/bsd-license.php
 * @todo            Get this package from composer.
 */
abstract class Output
{
    /**
     * @var array Captured Headers which would have been sent to the output buffer.
     */
    public static array $headers = [];
    /**
     * @var string|null Captured body which would have been sent to the output buffer.
     */
    public static ?string $body;

    /**
     * Clear the captured output headers and body.
     *
     * @return void
     */
    public static function reset()
    {
        self::$headers = [];
        self::$body    = null;
        @ob_clean();
    }

    /**
     * Get the captured body.
     *
     * @return string|null
     */
    public static function getBody(): ?string
    {
        Output::$body .= ob_get_contents();
        ob_clean();

        return self::$body;
    }

    /**
     * Wrapper function to include function overrides.
     *
     * To override functions, the overrides need to be declared somewhere.
     * To keep this more versatile and tidy, overrides and their namespace should be declared in a separate php file.
     *
     * @param   string  $functionsFile  Path of php file which contains the function overrides.
     *
     * @return void
     */
    public static function getOverrides(string $functionsFile): void
    {
        require $functionsFile;
    }
}


updateChangelog.php
<?php

/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2022, Ferry Cools (DigiLive)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

// Instantiate composer's auto loader.

require __DIR__ . '/vendor/autoload.php';

// My options.
$release         = [
    'determined' => false,
    'version'    => 'v2.0.0',
    'date'       => '2022-10-02',
];
$logType         = 'MarkDown';
$gitProvider     = 'github';
$tagOrder        = 'desc';
$titleOrder      = 'asc';
$changelogLabels = ['Add', 'Cut', 'Fix', 'Bump', 'Optimize'];

// Default settings.
$changelogSettings = [
    'headTagName' => 'Next Release',
    'headTagDate' => 'Soon',
    'tagOrder'    => $tagOrder,
    'titleOrder'  => $titleOrder,
];

// Settings for determined release.
if ($release['determined']) {
    $changelogSettings['headTagName'] = $release['version'];
    $changelogSettings['headTagDate'] = $release['date'];
}

// Git provider settings.
$github = [
    'urls'     => [
        'commit'       => 'https://github.com/DigiLive/fileStreamer/commit/{commit}',
        'issue'        => 'https://github.com/DigiLive/fileStreamer/issue/{issue}',
        // No merge-request url because GitHub does not make a difference between issues and pull-requests.
        'mergeRequest' => null,
    ],
    'patterns' => [
        'issue'        => '#(\d+)',
        // No merge-request pattern because GitHub does not make a difference between issues and pull-requests.
        'mergeRequest' => null,
    ],
];

// Setup changelog.
$className = "DigiLive\\GitChangelog\\Renderers\\$logType";
$changelog = new $className();

$changelog->setUrl('commit', $$gitProvider['urls']['commit']);
$changelog->setUrl('issue', $$gitProvider['urls']['issue']);
$changelog->setUrl('mergeRequest', $$gitProvider['urls']['mergeRequest']);

$changelog->setPattern('issue', $$gitProvider['patterns']['issue']);
$changelog->setPattern('mergeRequest', $$gitProvider['patterns']['mergeRequest']);

$changelog->setLabels(...$changelogLabels);

$changelog->setOptions($changelogSettings);

//$changelog->setFromTag('v0.1.0');
//$changelog->setToTag('v0.1.0');

// Create changelog.
try {
    $start = microtime(true);
    $changelog->build();
    $fileName          = 'MarkDown' == $logType ? 'CHANGELOG.md' : 'CHANGELOG.html';
    $time_elapsed_secs = microtime(true) - $start;
    echo "Building Time: $time_elapsed_secs seconds\n";
    file_put_contents($fileName, $changelog->get());
} catch (Exception $e) {
    echo $e->getMessage();
}


src/FileStreamer.php
<?php

declare(strict_types=1);

namespace DigiLive\FileStreamer;

use finfo;
use RuntimeException;
use SplFileInfo;

/**
 * FileStreamer
 *
 * A library for streaming a local file to a client.
 * Supports inline and attachment disposition, single file download and resumable, single/multiple range downloads.
 *
 * PHP version 7.4 or greater
 *
 * @package         DigiLive\FileStreamer
 * @author          Ferry Cools <info@DigiLive.nl>
 * @copyright   (c) 2022 Ferry Cools
 * @license         New BSD License http://www.opensource.org/licenses/bsd-license.php
 * @version         1.0.0
 * @link            https://github.com/DigiLive/fileStreamer
 */
class FileStreamer
{
    private const CONTENT_INLINE     = 0;
    private const CONTENT_ATTACHMENT = 1;
    private const CONTENT_PARTIAL    = 2;
    private const RANGE_INVALID      = 4;

    /**
     * @var bool True for inline disposition. False for attachment disposition.
     */
    private bool $inline = false;
    /**
     * @var SplFileInfo Details of the requested file.
     * @see SplFileInfo
     */
    private SplFileInfo $fileInfo;
    /**
     * @var int Delay in ms to slow down sending the file content to the client.
     *          Increase this value when serving the file hogs up the systems resources.
     */
    private int $delay;
    /**
     * @var array The ranges in bytes which are extracted from the http request headers.
     *            Each element contains a range indicated by a start- and an end byte.
     */
    private array $fileRanges;

    /**
     * @var array Contains the formats to create content for a multipart file download.
     *            A multipart file download is identified by having multiple byte-ranges in the http request headers.
     */
    private array $multipartFormats;
    /**
     * @var false|resource Handle to the file to serve.
     */
    private $filePointer;
    /**
     * @var string|null If not null, it defines type mimetype of the file to download.
     */
    private ?string $mimeType = null;

    /**
     * Class Constructor.
     *
     * Any trailing slash of the given path is stripped from the string parameter.
     * Optionally you can set a delay time to prevent hogging up the system resources.
     *
     * @param   string  $filePath  Path to the file to serve.
     * @param   int     $delay     Delay in micro second.
     */
    public function __construct(string $filePath, int $delay = 0)
    {
        $this->fileInfo = new SplFileInfo(trim(rtrim($filePath), '\/'));
        $this->delay    = $delay;
    }

    /**
     * Set the disposition of the file stream to inline.
     *
     * This way the library acts like a proxy server for the requested file.
     * A value of false will set the disposition to attachment.
     *
     * @param   bool  $inline  True to enable inline disposition.
     *
     * @return void
     */
    public function setInline(bool $inline = true): void
    {
        $this->inline = $inline;
    }

    /**
     * Set the mimetype of to file to serve.
     *
     * When not null, it overrides getting the mimetype automatically when the headers are sent.
     *
     * @param   string|null  $mimeType  the mimetype of the file to serve.
     *
     * @return void
     */
    public function setMimeType(?string $mimeType): void
    {
        $this->mimeType = $mimeType;
    }

    /**
     * Start serving the file.
     *
     * @return void
     */
    public function start()
    {
        $fileName = $this->fileInfo->getFilename();

        $this->filePointer = @fopen($this->fileInfo->getPathname(), 'rb');
        if (!$this->filePointer || !flock($this->filePointer, LOCK_SH | LOCK_NB)) {
            throw new RuntimeException("File $fileName is currently not available!");
        }

        $this->disableCompression();
        $this->getRequestedRanges();

        switch (count($this->fileRanges)) {
            case 0:
                $this->sendFile();
                break;
            case 1:
                $this->sendSingleRange();
                break;
            default:
                $this->sendMultipleRanges();
        }

        // Serving the file finished successfully.
        if (@fclose($this->filePointer) === false) {
            trigger_error("An error occurred while closing file $fileName!", E_USER_WARNING);
        }

        $this->terminate('File served');
    }

    /**
     * Disable output compression of the servers.
     *
     * Note:
     * Browser compression can only be disabled when php runs as an apache module. For other php modes, you'll
     * need to disable the compression in your server configuration or .htaccess file.
     *
     * @return void
     */
    private function disableCompression(): void
    {
        if (PHP_SAPI == 'apache2handler' && @apache_setenv('no-gzip', '1') === false) {
            trigger_error('An error occurred while disabling output compression of the webserver!', E_USER_WARNING);
        }
        if (@ini_set('zlib.output_compression', 'Off') === false) {
            // PHP Unit test will always trigger this error. Suppress with @ while testing.
            trigger_error('An error occurred while disabling output compression of the php server!', E_USER_WARNING);
        }
    }

    /**
     * Get the requested ranges from the headers of the request.
     *
     * One or multiple ranges are extracted and sanitized from the request header.
     * - A range start is always lte 0 and end.
     * - A range end is always lt the filesize.
     *
     * If these conditions are not met after sanitation, the script sends a http 416 error and stops execution.
     */
    private function getRequestedRanges(): void
    {
        /*
         * Valid ranges are:
         * bytes=0-500                  // The first 500 bytes.
         * bytes=-500                   // The last 500 bytes, not 0-500!
         * bytes=500-                   // From byte 500 tot the end.
         * bytes=0-500,1000-1499,-200   // The first 500 bytes, From byte 1000 to 1499 and the last 200 bytes.
         */

        if (!isset($_SERVER['HTTP_RANGE'])) {
            $this->fileRanges = [];

            return;
        }

        $fileRanges = [];
        $fileEnd    = $this->fileInfo->getSize() - 1;
        [$rangeUnit, $requestedRanges] = explode('=', $_SERVER['HTTP_RANGE'], 2);

        if ($rangeUnit != 'bytes') {
            $this->sendHeaders(self::RANGE_INVALID);
        }

        // Sanitize requested ranges.
        $requestedRanges = explode(',', $requestedRanges);
        foreach ($requestedRanges as $range) {
            [$start, $end] = explode('-', $range);

            if ($start == '') {
                // bytes=-500 The last 500 bytes, not 0-500!
                $start = $fileEnd - $end + 1;
                $end   = $fileEnd;
            }

            if ($end == '') {
                // bytes=500- From byte 500 to the end.
                $end = $fileEnd;
            }

            $start = (int)max($start, 0);
            $end   = (int)min($end, $fileEnd);

            if ($start > $end) {
                $this->sendHeaders(self::RANGE_INVALID);
            }

            $fileRanges[] = [
                'start' => $start,
                'end'   => $end,
            ];
        }

        $this->fileRanges = $fileRanges;
    }

    /**
     * Send appropriate raw HTTP headers.
     *
     * Remember that headers must be sent before any actual output is sent, either by normal HTML tags, blank lines in
     * a file, or from PHP.
     *
     * @param   int  $type  Type of headers to send.
     *
     * @return void
     */
    private function sendHeaders(int $type): void
    {
        if ($type == self::RANGE_INVALID) {
            header('HTTP/1.1 416 Requested Range Not Satisfiable');
            $this->terminate();
        }

        // Get file mimetype.
        $filePath = $this->fileInfo->getPathname();
        $fileSize = $this->fileInfo->getSize();
        $mimeType = $this->mimeType;
        if (null === $mimeType) {
            $fileInfo = new finfo();
            $mimeType = @$fileInfo->file($filePath, FILEINFO_MIME_TYPE);
        }

        // Caching headers as IE6 workaround.
        header('Pragma: public');
        header('Expires: -1');
        header('Cache-Control: public, must-revalidate, post-check=0, pre-check=0');

        // Range header.
        header('Accept-Ranges: bytes');

        // Content headers.
        $contentType = 'Content-Type: ' . $mimeType ?: 'application/octet-stream';
        header($contentType);
        header('Content-Transfer-Encoding: binary');
        header("Content-Disposition: attachment; filename=\"{$this->fileInfo->getFilename()}\"");
        header("Content-Length: $fileSize");

        switch ($type) {
            case self::CONTENT_INLINE:
                header('Content-Disposition: inline');
                break;
            case self::CONTENT_PARTIAL:
                header('HTTP/1.1 206 Partial Content');

                $rangeCount = count($this->fileRanges);

                // Single range.
                if ($rangeCount == 1) {
                    header('Content-Length: ' . ($this->fileRanges[0]['end'] - $this->fileRanges[0]['start'] + 1));
                    header(
                        sprintf(
                            'Content-Range: bytes %d-%d/%d',
                            $this->fileRanges[0]['start'],
                            $this->fileRanges[0]['end'],
                            $fileSize
                        )
                    );

                    return;
                }

                // Multiple ranges.
                $contentLength = $rangeCount * 38;                                   // boundaryStart
                $contentLength += $rangeCount * strlen($contentType . "\r\n");       // contentType
                $contentLength += 40;                                                // boundaryEnd

                // Calculate the content length of the parted download.
                foreach ($this->fileRanges as $range) {
                    $contentLength += strlen(
                        sprintf(
                            $this->multipartFormats['rangeFormat'],
                            $range['start'],
                            $range['end'],
                            $fileSize
                        )
                    );
                    $contentLength += $range['end'] - $range['start'] + 1;
                }

                header("Content-Length: $contentLength");
                header('Content-Type: multipart/byteranges; boundary=' . $this->multipartFormats['rangeBoundary']);
        }
    }

    /**
     * Terminate the current script with exit message/code.
     *
     * Terminates execution of the script. Shutdown functions and object destructors will always be executed.
     * If status is a string, this function prints the status just before exiting.
     * If status is an int, that value will be used as the exit status and not printed. Exit statuses should be in the
     * range 0 to 254, the exit status 255 is reserved by PHP and shall not be used. The status 0 is used to terminate
     * the program successfully.
     *
     * @param   $status
     *
     * @return void
     */
    public function terminate($status = null): void
    {
        exit($status);
    }

    /**
     * Stream the complete file to the client.
     *
     * @return void
     */
    private function sendFile()
    {
        $this->sendHeaders($this->inline ? self::CONTENT_INLINE : self::CONTENT_ATTACHMENT);
        $this->flush(0, $this->fileInfo->getSize());
    }

    /**
     * Flush buffered content of the file to the client.
     *
     * The start and end of the content are defined by the methods parameters.
     * The values of the parameters are treated as number of bytes from the beginning of the file to serve.
     *
     * @param   int  $start  Start of the content.
     * @param   int  $end    End of the content.
     *
     * @return void
     */
    private function flush(int $start, int $end): void
    {
        $done = false;

        fseek($this->filePointer, $start);

        while (!$done && connection_status() == CONNECTION_NORMAL) {
            ini_set('max_execution_time', '30');
            echo @fread($this->filePointer, min(1024, $end - $start + 1));
            ob_flush();
            flush();
            $done = feof($this->filePointer) || ftell($this->filePointer) >= $end;
            usleep($this->delay);
        }
    }

    /**
     * Send a single range of file content to the client.
     *
     * @return void
     */
    private function sendSingleRange(): void
    {
        $this->sendHeaders(self::CONTENT_PARTIAL);
        $this->flush($this->fileRanges[0]['start'], $this->fileRanges[0]['end']);
    }

    /**
     * Send multiple ranges of file content to the client.
     *
     * @return void
     */
    private function sendMultipleRanges(): void
    {
        $fileInfo = new finfo();
        $mimeType = @$fileInfo->file($this->fileInfo->getPathname(), FILEINFO_MIME_TYPE);
        $fileSize = $this->fileInfo->getSize();

        $this->setMultipartFormats();
        $this->sendHeaders(self::CONTENT_PARTIAL);

        foreach ($this->fileRanges as $range) {
            echo $this->multipartFormats['boundaryStart'];
            echo 'Content-Type: ' . ($mimeType ?: 'application/octet-stream') . "\r\n";
            echo sprintf($this->multipartFormats['rangeFormat'], $range['start'], $range['end'], $fileSize);
            $this->flush($range['start'], $range['end']);
        }
        echo $this->multipartFormats['boundaryEnd'];
    }

    /**
     * Create the boundary and range strings for multipart downloads.
     *
     * @return void
     */
    private function setMultipartFormats(): void
    {
        $rangeBoundary          = md5($this->fileInfo->getPathname());
        $this->multipartFormats = [
            'rangeBoundary' => "$rangeBoundary",
            'boundaryStart' => "\r\n--$rangeBoundary\r\n",
            'boundaryEnd'   => "\r\n--$rangeBoundary--\r\n",
            'rangeFormat'   => "Content-range: bytes %d-%d/%d\r\n\r\n",
        ];
    }
}
